src/lib/statistics/statistics_objects/top_k_as_generic_histogram.cpp

```diff
@@ -0,0 +1,129 @@
+                   top_k_front_value_distribution.end(),
+                   [&](const auto& left_value_count, const auto& right_value_count) {
+                     return left_value_count.second > right_value_count.second;
+                   });
+
+  // Sort Top K values lexicographically
+  // We later use the lexicographically sorted Top K values for easier and more performant histogram construction.
+  std::sort(top_k_front_value_distribution.begin(), top_k_front_value_distribution.begin() + k,
+            [&](const auto& left_value_count, const auto& right_value_count) {
+              return left_value_count.first < right_value_count.first;
+            });
+
+  for (auto top_k_index = 0u; top_k_index < k; top_k_index++) {
+    top_k_names[top_k_index] = top_k_front_value_distribution[top_k_index].first;
+    top_k_counts[top_k_index] = top_k_front_value_distribution[top_k_index].second;
+  }
+
+  // Remove Top K values from value distribution
+  for (auto top_k_index = 0u; top_k_index < k; top_k_index++) {
+    auto value_distribution_it = remove(value_distribution.begin(), value_distribution.end(),
+                                        std::make_pair(top_k_names[top_k_index], top_k_counts[top_k_index]));
+    value_distribution.erase(value_distribution_it, value_distribution.end());
+  }
+
+  // Each Top K value is modeled as one bin with height as its stored count.
+  // Between two Top K value bins, one bin is created for the Non-Top K values between them.
+  // Together with a potential last bin after the last Top K value we have a maximum bin count of 2 * K + 1.
+  // If there are no more values stored in value_distribution after the Top K values have been removed,
+  // we only have Top K values and therefore need exactly k bins.
+  const auto bin_count = value_distribution.size() < 1 ? BinID{k} : BinID{2 * top_k_names.size() + 1};
+
+  GenericHistogramBuilder<T> builder{bin_count, domain};
+
+  // Calculate estimate for occurrence count for Non-Top K values assuming a uniform distribution.
+  const auto non_top_k_count =
+      std::accumulate(value_distribution.cbegin(), value_distribution.cend(), HistogramCountType{0},
+                      [](HistogramCountType current_count, const std::pair<T, HistogramCountType>& value_count) {
+                        return current_count + value_count.second;
+                      });
+
+  const auto bin_distinct_count = value_distribution.size();
+  const auto count_per_non_top_k_value =
+      bin_distinct_count != 0 ? non_top_k_count / static_cast<float>(bin_distinct_count) : BinID{0};
+
+  // Construct Generic Histogram with single bins for Top K Values.
+  // For Non-Top K Values, one bin is created for all Non-Top K values between two Top K bins,
+  // using the calculated estimation of count_per_non_top_k_value.
+
+  auto current_minimum_index = 0u;
+  auto current_maximum_index = value_distribution.size() - 1;
```

Comment:
First, it looks like this could underflow. Also, isn't this overwritten before use in line 102?

Reply:
We wanted to make use of auto to initialise the `current_maximum_index`, would only declaring the `current_maximum_index` variable as an `uint` be preferable?

Reply:
I see. Still, this is confusing, I would probably go for a `size_t` then.